core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -441,11 +444,17 @@ public SqlNode toSql(RexProgram program, RexNode rex) {
         return field(((RexInputRef) rex).getIndex());
 
       case FIELD_ACCESS:
-        RexFieldAccess access = (RexFieldAccess) rex;
-        final RexCorrelVariable variable =
-            (RexCorrelVariable) access.getReferenceExpr();
-        final Context aliasContext = correlTableMap.get(variable.id);
-        return aliasContext.field(access.getField().getIndex());
+        final RexFieldAccess access = (RexFieldAccess) rex;
+        switch (access.getReferenceExpr().getKind()) {
+        case CORREL_VARIABLE:
+          final RexCorrelVariable variable = (RexCorrelVariable) access.getReferenceExpr();
+          final Context aliasContext = correlTableMap.get(variable.id);
+          return aliasContext.field(access.getField().getIndex());
+        default:
```

Comment:
The code assumes that FIELD_ACCESS wraps always a CORREL_VARIABLE or INPUT_REF. I have the impression that this code will lead again to a ClassCastException just a bit later (e.g., for a doubly nested struct where there is a RexFieldAccess(RexFieldAccess(RexInputRef)).

Reply:
You're right, I thought that it could only go down one level but after adding tests for multiple nested levels I see that it can happen, I am updating the PR to support this and add tests for this case